Native scrollframe

boatbomber opened the pull request:
The custom scrollingframes are cool, but it's a bad move. Trust me, I've been down this road.

Back before ScrollingFrames were smooth scrolling, I made a custom one for that. It was a pain to maintain, so instead I reworked it into [SmoothScroll](https://devforum.roblox.com/t/smoothscroll-module-superseded-by-roblox-update-421/313619) which added smooth functionality to default ScrollingFrames.

The main problems with using custom are:
- Maintaining
- Input methods (touchpad, gamepad, touchscreen)
- OS level input settings we cant read
- Input sinking

If you have nested scrollframes our custom method does not handle those correctly.

I've rewritten the scrollframe component to essentially just be a stylized bar overlay on top of an normal scrollingframe so all that stuff is correctly handled by the engine- all we really care about here is the QT styled scrollbar anyway, we aren't adding custom functionality (beyond the up/down buttons, but I've included those in this version too).

![image](https://user-images.githubusercontent.com/40185666/162535692-40643a0c-68fa-407c-9a4f-e49ea39011e7.png) ![image](https://user-images.githubusercontent.com/40185666/162535731-df72f63c-7927-48af-a80f-8a8423dee06f.png)

boatbomber committed:
Add project structure

boatbomber committed:
Fix Hydrate typo

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'mvyasu:main' into main

boatbomber committed:
Merge branch 'main' of https://github.com/boatbomber/PluginEssentials into main

boatbomber committed:
Rewrite scrollframe from scratch to use native ScrollingFrame instances

boatbomber committed:
Merge branch 'main' of https://github.com/mvyasu/PluginEssentials into main

boatbomber committed:
Merge branch 'main' of https://github.com/boatbomber/PluginEssentials into native-scrollframe

boatbomber commented:
I've created a draft for you to give early feedback. I know I still need to add the color themes and give instances clear names.

Edit: Forgot that I actually did add colors already, whoops. Just the names and then it's ready for review.

boatbomber committed:
Name instances clearly

mvyasu commented on src/StudioComponents/ScrollFrame/init.lua:
```diff
@@ -1,467 +1,271 @@
-	ScrollingDirection: (Enum.ScrollingDirection | types.StateObject<Enum.ScrollingDirection>)?,
-	UIPadding: UIPadding?,
-	Layout: UILayout?,
+	ScrollingEnabled: (boolean | types.StateObject<boolean>)?,
```
I think it'd be good to keep consistency with the Enabled variable named. I'm sure people would assume/infer that if the ScrollFrame is not enabled, they couldn't scroll.

boatbomber replied:
I'm on the fence about that one, since scrollingframes have ScrollingEnabled as the property not Enabled.

mvyasu replied:
If it's an actual property for ScrollingFrames, then you can keep it. I've never heard of it.

mvyasu replied:
However, you should still get rid of the `Enabled` from the properties type if that's case.

mvyasu commented on src/StudioComponents/ScrollFrame/init.lua:
```diff
@@ -1,467 +1,271 @@
-	Layout: UILayout?,
+	ScrollingEnabled: (boolean | types.StateObject<boolean>)?,
+	VerticalScrollBarPosition: (EnumItem | types.StateObject<EnumItem>)?,
+	VerticalScrollBarInset: (EnumItem | types.StateObject<EnumItem>)?,
```
should use `Enum.ScrollBarInset` instead of being vague with `EnumItem`. The same goes for the `EnumItem` type above, except it'd be `Enum.VerticalScrollBarPosition`.

boatbomber committed:
Fix types

mvyasu commented:
Some things that'll need to be done before this is merged with the main branch:

- `ScrollFrame`'s should not be transparent and should have a default size of `UDim2.fromScale(1, 1)`.
- Passing `BackgroundColor3` to `ScrollFrame` will override the colors for the handles and bar as well or there should be separate properties to change their colors
- `ScrollFrame` is a base component that can do both horizontal and vertical scrolling with scroll bars
- `LayoutScrollFrame` is a component that uses `ScrollFrame` but handles the `CanvasSize` for both `UIPadding` and a `UILayout`

mvyasu committed:
ZIndexBehavior Global Support

* Ensured that descendants of ScrollFrame, like the scroll bars, should still be visible even if ZIndexBehavior is set to Global. This is what all the other components I wrote do

mvyasu committed:
Improved native scrollframe

- Added a getModifier util module
- Renamed ScrollFrame to BaseScrollFrame
- Added a new component called ScrollFrame that handles the CanvasSize of the BaseScrollFrame with the provided props
- Removed layout-related references from BaseScrollFrame
- Removed the Constants module from the BaseScrollFrame component
- Added a ScrollBar module that the BaseScrollFrame uses
- Improved the ScrollArrow module to be more flexible
- The new ScrollBar that I've put together has the handle highlight and the BorderMode can be changed
- The new BaseScrollFrame should match more with the Roblox's
- Numerous small changes that should make it easier to modify the BaseScrollFrame component
